File: Topological/207_CourseSchedule.py
from collections import deque
from typing import List

"""
207. Course Schedule
Difficulty -> Medium

This is THE topological course question basically, we see all the incoming
edges of each node, then we start with the node with zero incoming edges
then removing what they point to specifically, until we have reached all
nodes, then will terminate if there is a cycle or not.

The time complexity is O(V + E) as we iterate through all known Edges E and
nodes (vertices) V. This will technically be Theta(V+E) because in the best
case we also run it in V+E time. 
"""

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        if not prerequisites:
            return True

        # add all courses and all of their dependencies
        courseMap = {i: [] for i in range(numCourses)}
        inMap = {i: 0 for i in range(numCourses)}
        
        for post, pre in prerequisites:
            courseMap[pre].append(post)
            inMap[post] += 1
            
        q = deque([u for u in inMap if inMap[u] == 0])
        finalOrder = set()

        while q:
            curr = q.popleft()
            finalOrder.add(curr)

            for neighbor in courseMap[curr]:
                inMap[neighbor] -= 1
                if inMap[neighbor] == 0:
                    q.append(neighbor)

        return len(finalOrder) == numCourses


"""
RETRY:

After not being on leetcode for a millenium it feels like, I am back and I have
attempted this problem again... And I shit you not it looks the EXACT SAME

The only difference I want to point out is that my variable names this time around
look significantly better than before, thank you.
"""


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        if not prerequisites:
            return True

        # make a dictionary for the number of prereqs left for this course
        numPrereq = {i: 0 for i in range(numCourses)}
        # make a dictionary for all the classes this one is a prereq for
        nextCourses = {i: [] for i in range(numCourses)}

        for post, pre in prerequisites:
            # add a prereq and add it to the list for pre
            numPrereq[post] += 1
            nextCourses[pre].append(post)

        q = deque([i for i in numPrereq if numPrereq[i] == 0])
        finalOrder = set()
        
        while q:
            value = q.popleft()
            finalOrder.add(value)

            for nextClass in nextCourses[value]:
                numPrereq[nextClass] -= 1
                if numPrereq[nextClass] == 0:
                    q.append(nextClass)
        
        return len(finalOrder) == numCourses

File: Topological/207_CourseSchedule.cpp
#include "../libraries.h"

using namespace std;

/**
 * 207. Course Schedule 
 * 
 * The classic topological sort problem
 * This one is easy to understand but slightly confusing to understand
 * in my opinion
 * 
 * Otherwise I am happy I finally did it in C++
 */
class Solution {
    public:
        bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
            if (prerequisites.empty()) { return true; }

            unordered_map<int, int> prereqs;
            unordered_map<int, vector<int> > nextCourses;

            // add all classes taken to the maps
            for (int cl = 0; cl < numCourses; cl++) {
                prereqs[cl] = 0;
                vector<int> nextCl;
                nextCourses[cl] = nextCl;
            }
    
            for (int i = 0; i < prerequisites.size(); i++) {
                // get the next class and the class you need to take it
                int nextClass = prerequisites[i][0], preClass = prerequisites[i][1];
                // add the next class to the prereq
                prereqs[nextClass] += 1;
                // add the next course to the pre classes list
                nextCourses[preClass].push_back(nextClass);
            }
    
            // make a queue and add all prereqs to it
            queue<int> q;
            for (const auto& kv : prereqs) {
                if (kv.second == 0) { q.push(kv.first); }
            }
    
            // go through the queue and add all the next values
            int count = 0;
            while(q.size() > 0) {
                int curr = q.front();
                q.pop();
    
                count += 1;
    
                for (int newClass : nextCourses[curr]) {
                    prereqs[newClass] -= 1;
                    if (prereqs[newClass] == 0) { q.push(newClass); }
                }
            }
            return count == numCourses;
        }
};